stdlib/std.jsonnet

```diff
@@ -88,6 +88,41 @@ limitations under the License.
     range(from, to)::
         std.makeArray(to - from + 1, function(i) i + from),
 
+    slice(indexable, index, end, step)::
+        if (index != null && index < 0) || (end != null && end < 0) || (step != null && step < 0) then
+            error("got [%s:%s:%s] but negative index, end, and steps are not implemented" % [index, end, step] )
+        else if std.type(indexable) != "string" && std.type(indexable) != "array" then
+            error("std.slice accepts a string or an array, but got: %s" % std.type(indexable))
+        else
+            local invar =
+                // loop invariant with defaults applied
+                {
+                    indexable: indexable,
+                    index:
+                        if index == null then 0
+                        else index,
+                    end:
+                        if end == null then std.length(indexable)
+                        else end,
+                    step:
+                        if step == null then 1
+                        else step,
+                    length: std.length(indexable),
+                    type: std.type(indexable)
+                };
+            local build(slice, cur) =
+                if cur >= invar.end || cur >= invar.length then
+                    slice
+                else
+                    build(
+                        if invar.type == "string" then
+                            slice + invar.indexable[cur]
+                        else
+                            slice + [ invar.indexable[cur] ],
+                        cur + invar.step
+                    ) tailstrict;
```

Comment:
O(n^2) :( but would rather optimize this in the vm for array concatinations where one side is not used again.

```console
$ jsonnet git:(slice): cat exp.jsonnet 
std.range(0, std.extVar("vars").count)[2:100000000000:2]
$ jsonnet git:(slice): time ./jsonnet --code-var vars='{"count": 1000}' exp.jsonnet > /dev/null 
./jsonnet --code-var vars='{"count": 1000}' exp.jsonnet > /dev/null  0.25s user 0.00s system 99% cpu 0.258 total
$ jsonnet git:(slice): time ./jsonnet --code-var vars='{"count": 10000}' exp.jsonnet > /dev/null                                                                                             
./jsonnet --code-var vars='{"count": 10000}' exp.jsonnet > /dev/null  1.98s user 0.01s system 99% cpu 1.990 total
$ jsonnet git:(slice): time ./jsonnet --code-var vars='{"count": 100000}' exp.jsonnet > /dev/null
./jsonnet --code-var vars='{"count": 100000}' exp.jsonnet > /dev/null  46.23s user 0.30s system 99% cpu 46.594 total
```